File: codewave_py/command.py
import codewave_core.storage as storage
import codewave_core.logger as logger
import codewave_core.codewave

def _optKey(key,dict,defVal = None): 
	# optional Dictionary key
	return dict[key] if key in dict else defVal

class InitialiserSet(set):
	def add(self,initialiser):
		existing = self.exists(initialiser)
		if existing :
			self.remove(existing)
		super(InitialiserSet, self).add(initialiser)
	def exists(self,initialiser):
		print(initialiser.__module__ + initialiser.__name__)
		for init in self:
			if init.__module__ == initialiser.__module__ and init.__name__ == initialiser.__name__ :
				return True
		return False
	
if 'cmdInitialisers' not in vars() :
	cmdInitialisers = InitialiserSet()
	cmdIniters = cmdInitialisers
	
def initCmds():
	global cmds
	global cmdInitialisers
	cmds = Command(None,{
		'cmds':{
			'hello':'Hello, World!'
		}
	})
	for initialiser in cmdInitialisers:
		initialiser()

def saveCmd(fullname,data):
	global cmds
	cmds.setCmd(fullname,Command(fullname.split(':').pop(),data))
	savedCmds = storage.load('cmds')
	if savedCmds is None :
		savedCmds = {}
	savedCmds[fullname] = data
	storage.save('cmds',savedCmds)

def loadCmds():
	global cmds
	savedCmds = storage.load('cmds')
	if savedCmds is not None :
		for fullname, data in savedCmds.items():
			cmds.setCmd(fullname,Command(fullname.split(':').pop(),data))

	
class Command():
	def __init__(self,name,data=None,parent=None):
		self.name,self.data = name,data
		self.cmds = []
		self.detectors = []
		self.executeFunct = self.resultFunct = self.resultStr = self.aliasOf = self.cls = None
		self.aliased = None
		self.fullName = self.name
		self.depth = 0
		self._parent, self._inited = None, False
		self.setParent(parent)
		self.defaults = {}
		
		self.defaultOptions = {
			'nameToParam' : None,
			'checkCarret' : True,
			'parse' : False,
		}
		self.options = {}
		self.finalOptions = None
	@property
	def parent(self):
			return self._parent
	@parent.setter
	def parent(self, value):
		self.setParent(value)
	def setParent(self, value):
		if self._parent != value :
			self._parent = value
			self.fullName = (
				self._parent.fullName + ':' + self.name 
				if self._parent is not None and self._parent.name is not None 
				else self.name
			)
			self.depth = (
				self._parent.depth + 1
				if self._parent is not None
				else 0
			)
	def init(self):
		if not self._inited :
			self._inited = True
			self.parseData(self.data)
		return self
	def isEditable(self):
		return self.resultStr is not None
	def isExecutable(self):
		for p in ['resultStr','resultFunct','aliasOf','cls','executeFunct'] :
			if getattr(self, p) is not None:
				return True
		return False
	def resultIsAvailable(self,instance = None):
		if instance is not None and instance.cmdObj is not None:
			return instance.cmdObj.resultIsAvailable()
		for p in ['resultStr','resultFunct'] :
			if getattr(self, p) is not None:
				return True
		return False
	def getDefaults(self,instance = None):
		res = {}
		aliased = self.getAliased(instance)
		if aliased is not None :
			res.update(aliased.getDefaults(instance))
		res.update(self.defaults)
		if instance is not None and instance.cmdObj is not None:
			res.update(instance.cmdObj.getDefaults())
		return res;
	def result(self,instance):
		if instance.cmdObj is not None:
			return instance.cmdObj.result()
		aliased = self.getAliased(instance)
		if aliased is not None :
			return aliased.result(instance)
		if self.resultFunct is not None:
			return self.resultFunct(instance)
		if self.resultStr is not None:
			return self.resultStr
	def execute(self,instance):
		if instance.cmdObj is not None:
			return instance.cmdObj.execute()
		aliased = self.getAliased(instance)
		if aliased is not None :
			return aliased.execute(instance)
		if self.executeFunct is not None:
			return self.executeFunct(instance)
	def getExecutableObj(self,instance):
		self.init()
		if self.cls is not None :
			return self.cls(instance)
		aliased = self.getAliased(instance)
		if aliased is not None :
			return aliased.getExecutableObj(instance)
	def getAliased(self,instance = None):
		if instance is not None and instance.cmd == self and instance.aliasedCmd is not None :
			return instance.aliasedCmd or None
		if self.aliasOf is not None :
			if instance is None :
				codewave = codewave_core.codewave.Codewave()
			else :
				codewave = instance.codewave
			aliased = codewave.getCmd(self.aliasOf)
			if instance is not None :
				instance.aliasedCmd = aliased or False
			return aliased
	def setOptions(self,data):
		for key, val in data.items():
			if key in self.defaultOptions:
				self.options[key] = val
	def getOptions(self,instance = None):
		if instance is not None and instance.cmdOptions is not None :
			return instance.cmdOptions
			
		opt = {}
		opt.update(self.defaultOptions)
		aliased = self.getAliased(instance)
		if aliased is not None :
			opt.update(aliased.getOptions(instance))
		opt.update(self.options)
		if instance is not None and instance.cmdObj is not None:
			opt.update(instance.cmdObj.getOptions())
		if instance is not None :
			instance.cmdOptions = opt
		return opt
	def getOption(self,key,instance = None):
		options = self.getOptions(instance)
		if key in options:
			return options[key]
	def parseData(self,data):
		self.data = data
		if isinstance(data, str):
			self.resultStr = data
			return True
		elif isinstance(data,dict) :
			return self.parseDictData(data)
		return False
	def parseDictData(self,data):
		res = _optKey('result',data)
		if hasattr(res, '__call__') :
			self.resultFunct = res
		elif res is not None :
			self.resultStr = res
			self.options['parse'] = True
		execute = _optKey('execute',data)
		if hasattr(execute, '__call__') :
			self.executeFunct = execute
		self.aliasOf = _optKey('aliasOf',data)
		self.cls = _optKey('cls',data)
		self.defaults = _optKey('defaults',data,self.defaults)
		
		self.setOptions(data)
		
		if 'help' in data :
			self.addCmd(self,Command('help',data['help'],self))
		if 'fallback' in data :
			self.addCmd(self,Command('fallback',data['fallback'],self))
		if 'cmds' in data :
			self.addCmds(data['cmds'])
		return True
	def addCmds(self,cmds):
		for name, data in cmds.items() :
			self.addCmd(Command(name,data,self))
	def addCmd(self,cmd):
		exists = self.getCmd(cmd.name)
		if exists is not None :
			self.removeCmd(exists)
			# exists.name = 'super'
			# cmd.addCmd(exists)
		cmd.setParent(self)
		self.cmds.append(cmd)
		return cmd
	def removeCmd(self,cmd):
		self.cmds.remove(cmd)
		return cmd
	def getCmd(self,fullname):
		self.init()
		parts = fullname.split(':',1)
		name = parts.pop()
		if len(parts) > 0 :
			return self.getCmd(parts[0]).getCmd(name)
		for cmd in self.cmds:
			if cmd.name == name:
				return cmd
	def setCmd(self,fullname,cmd):
		parts = fullname.split(':',1)
		name = parts.pop()
		if len(parts) > 0 :
			next = self.getCmd(parts[0])
			if next is None :
				next = self.addCmd(Command(parts[0]))
			return next.setCmd(name,cmd)
		else:
			self.addCmd(cmd)
			return cmd
	def addDetector(self,detector):
		self.detectors.append(detector)
			
	
class BaseCommand():
	def __init__(self,instance):
		self.instance = instance
	def resultIsAvailable(self):
		return hasattr(self,"result")
	def getDefaults(self):
		return {}
	def getOptions(self):
		return {}
				
				

File: lib/Command.js


const Context = require("./Context");

const Storage = require("./Storage").Storage;

const NamespaceHelper = require("./helpers/NamespaceHelper").NamespaceHelper;

var _optKey;

_optKey = function (key, dict, defVal = null) {
  // optional Dictionary key
  if (key in dict) {
    return dict[key];
  } else {
    return defVal;
  }
};

var Command = function () {
  class Command {
    constructor(name1, data1 = null, parent = null) {
      this.name = name1;
      this.data = data1;
      this.cmds = [];
      this.detectors = [];
      this.executeFunct = this.resultFunct = this.resultStr = this.aliasOf = this.cls = null;
      this.aliased = null;
      this.fullName = this.name;
      this.depth = 0;
      [this._parent, this._inited] = [null, false];
      this.setParent(parent);
      this.defaults = {};
      this.defaultOptions = {
        nameToParam: null,
        checkCarret: true,
        parse: false,
        beforeExecute: null,
        alterResult: null,
        preventParseAll: false,
        replaceBox: false,
        allowedNamed: null
      };
      this.options = {};
      this.finalOptions = null;
    }

    parent() {
      return this._parent;
    }

    setParent(value) {
      if (this._parent !== value) {
        this._parent = value;
        this.fullName = this._parent != null && this._parent.name != null ? this._parent.fullName + ':' + this.name : this.name;
        return this.depth = this._parent != null && this._parent.depth != null ? this._parent.depth + 1 : 0;
      }
    }

    init() {
      if (!this._inited) {
        this._inited = true;
        this.parseData(this.data);
      }

      return this;
    }

    unregister() {
      return this._parent.removeCmd(this);
    }

    isEditable() {
      return this.resultStr != null || this.aliasOf != null;
    }

    isExecutable() {
      var aliased, j, len, p, ref;
      aliased = this.getAliased();

      if (aliased != null) {
        return aliased.init().isExecutable();
      }

      ref = ['resultStr', 'resultFunct', 'cls', 'executeFunct'];

      for (j = 0, len = ref.length; j < len; j++) {
        p = ref[j];

        if (this[p] != null) {
          return true;
        }
      }

      return false;
    }

    isExecutableWithName(name) {
      var aliasOf, aliased, context;

      if (this.aliasOf != null) {
        context = new Context.Context();
        aliasOf = this.aliasOf.replace('%name%', name);
        aliased = this._aliasedFromFinder(context.getFinder(aliasOf));

        if (aliased != null) {
          return aliased.init().isExecutable();
        }

        return false;
      }

      return this.isExecutable();
    }

    resultIsAvailable() {
      var aliased, j, len, p, ref;
      aliased = this.getAliased();

      if (aliased != null) {
        return aliased.resultIsAvailable();
      }

      ref = ['resultStr', 'resultFunct'];

      for (j = 0, len = ref.length; j < len; j++) {
        p = ref[j];

        if (this[p] != null) {
          return true;
        }
      }

      return false;
    }

    getDefaults() {
      var aliased, res;
      res = {};
      aliased = this.getAliased();

      if (aliased != null) {
        res = Object.assign(res, aliased.getDefaults());
      }

      res = Object.assign(res, this.defaults);
      return res;
    }

    _aliasedFromFinder(finder) {
      finder.useFallbacks = false;
      finder.mustExecute = false;
      finder.useDetectors = false;
      return finder.find();
    }

    getAliased() {
      var context;

      if (this.aliasOf != null) {
        context = new Context.Context();
        return this._aliasedFromFinder(context.getFinder(this.aliasOf));
      }
    }

    getAliasedOrThis() {
      return this.getAliased() || this;
    }

    setOptions(data) {
      var key, results, val;
      results = [];

      for (key in data) {
        val = data[key];

        if (key in this.defaultOptions) {
          results.push(this.options[key] = val);
        } else {
          results.push(void 0);
        }
      }

      return results;
    }

    _optionsForAliased(aliased) {
      var opt;
      opt = {};
      opt = Object.assign(opt, this.defaultOptions);

      if (aliased != null) {
        opt = Object.assign(opt, aliased.getOptions());
      }

      return Object.assign(opt, this.options);
    }

    getOptions() {
      return this._optionsForAliased(this.getAliased());
    }

    getOption(key) {
      var options;
      options = this.getOptions();

      if (key in options) {
        return options[key];
      }
    }

    help() {
      var cmd;
      cmd = this.getCmd('help');

      if (cmd != null) {
        return cmd.init().resultStr;
      }
    }

    parseData(data) {
      this.data = data;

      if (typeof data === 'string') {
        this.resultStr = data;
        this.options['parse'] = true;
        return true;
      } else if (data != null) {
        return this.parseDictData(data);
      }

      return false;
    }

    parseDictData(data) {
      var execute, res;
      res = _optKey('result', data);

      if (typeof res === "function") {
        this.resultFunct = res;
      } else if (res != null) {
        this.resultStr = res;
        this.options['parse'] = true;
      }

      execute = _optKey('execute', data);

      if (typeof execute === "function") {
        this.executeFunct = execute;
      }

      this.aliasOf = _optKey('aliasOf', data);
      this.cls = _optKey('cls', data);
      this.defaults = _optKey('defaults', data, this.defaults);
      this.setOptions(data);

      if ('help' in data) {
        this.addCmd(new Command('help', data['help'], this));
      }

      if ('fallback' in data) {
        this.addCmd(new Command('fallback', data['fallback'], this));
      }

      if ('cmds' in data) {
        this.addCmds(data['cmds']);
      }

      return true;
    }

    addCmds(cmds) {
      var data, name, results;
      results = [];

      for (name in cmds) {
        data = cmds[name];
        results.push(this.addCmd(new Command(name, data, this)));
      }

      return results;
    }

    addCmd(cmd) {
      var exists;
      exists = this.getCmd(cmd.name);

      if (exists != null) {
        this.removeCmd(exists);
      }

      cmd.setParent(this);
      this.cmds.push(cmd);
      return cmd;
    }

    removeCmd(cmd) {
      var i;

      if ((i = this.cmds.indexOf(cmd)) > -1) {
        this.cmds.splice(i, 1);
      }

      return cmd;
    }

    getCmd(fullname) {
      var cmd, j, len, name, ref, ref1, space;
      this.init();
      [space, name] = NamespaceHelper.splitFirst(fullname);

      if (space != null) {
        return (ref = this.getCmd(space)) != null ? ref.getCmd(name) : void 0;
      }

      ref1 = this.cmds;

      for (j = 0, len = ref1.length; j < len; j++) {
        cmd = ref1[j];

        if (cmd.name === name) {
          return cmd;
        }
      }
    }

    setCmdData(fullname, data) {
      return this.setCmd(fullname, new Command(fullname.split(':').pop(), data));
    }

    setCmd(fullname, cmd) {
      var name, next, space;
      [space, name] = NamespaceHelper.splitFirst(fullname);

      if (space != null) {
        next = this.getCmd(space);

        if (next == null) {
          next = this.addCmd(new Command(space));
        }

        return next.setCmd(name, cmd);
      } else {
        this.addCmd(cmd);
        return cmd;
      }
    }

    addDetector(detector) {
      return this.detectors.push(detector);
    }

    static initCmds() {
      var j, len, provider, ref, results;
      Command.cmds = new Command(null, {
        'cmds': {
          'hello': {
            help: "\"Hello, world!\" is typically one of the simplest programs possible in\nmost programming languages, it is by tradition often (...) used to\nverify that a language or system is operating correctly -wikipedia",
            result: 'Hello, World!'
          }
        }
      });
      ref = this.providers;
      results = [];

      for (j = 0, len = ref.length; j < len; j++) {
        provider = ref[j];
        results.push(provider.register(Command.cmds));
      }

      return results;
    }

    static saveCmd(fullname, data) {
      return Promise.resolve().then(() => {
        return Command.cmds.setCmdData(fullname, data);
      }).then(() => {
        return this.storage.saveInPath('cmds', fullname, data);
      });
    }

    static loadCmds() {
      return Promise.resolve().then(() => {
        var savedCmds;
        return savedCmds = this.storage.load('cmds');
      }).then(savedCmds => {
        var data, fullname, results;

        if (savedCmds != null) {
          results = [];

          for (fullname in savedCmds) {
            data = savedCmds[fullname];
            results.push(Command.cmds.setCmdData(fullname, data));
          }

          return results;
        }
      });
    }

    static resetSaved() {
      return this.storage.save('cmds', {});
    }

    static makeVarCmd(name, base = {}) {
      base.execute = function (instance) {
        var p, val;
        val = (p = instance.getParam(0)) != null ? p : instance.content ? instance.content : void 0;

        if (val != null) {
          return instance.codewave.vars[name] = val;
        }
      };

      return base;
    }

    static makeBoolVarCmd(name, base = {}) {
      base.execute = function (instance) {
        var p, val;
        val = (p = instance.getParam(0)) != null ? p : instance.content ? instance.content : void 0;

        if (!(val != null && (val === '0' || val === 'false' || val === 'no'))) {
          return instance.codewave.vars[name] = true;
        }
      };

      return base;
    }

  }

  ;
  Command.providers = [];
  Command.storage = new Storage();
  return Command;
}.call(void 0);

exports.Command = Command;
var BaseCommand = class BaseCommand {
  constructor(instance1) {
    this.instance = instance1;
  }

  init() {}

  resultIsAvailable() {
    return this["result"] != null;
  }

  getDefaults() {
    return {};
  }

  getOptions() {
    return {};
  }

};
exports.BaseCommand = BaseCommand;

